misc: add a section to store kernel version information

SavinayDharmappa opened the pull request:
A section in linker is created to store the kernel version information .This feature is enabled based on CONFIG_VERSION_HEADER. This change make a provision to access the version information from the zephyr elf.

Signed-off-by: Savinay Dharmappa <savinay.dharmappa@intel.com>

galak commented:
No issue with this, but any particular reasons to add this?

nashif reviewed (changes requested):
Why is this only being done for x86? Can you please make this generic and available for everyone?

We also need a test that verifies this.

nashif commented on misc/Kconfig:
```diff
@@ -112,6 +112,13 @@ config CHECK_LINK_MAP
+	bool "Add a version header to binary"
+	default n
+	help
+	This will add a version header section to the generated binary.
```
2 spaces after tab

nvlsianpu replied:
Probably need to point to info on how-to provide content for the version header (processed by a post build script).

SavinayDharmappa replied:
ok. I will add it.

tyrantlizard77 reviewed (changes requested):
I'd like to understand why we need to do this.
Why does the version have to go in a special ELF section? Please describe what problem we are solving here.

ulfalizer commented on misc/Kconfig:
```diff
@@ -112,6 +112,13 @@ config CHECK_LINK_MAP
+
+config VERSION_HEADER
+	bool "Add a version header to binary"
+	default n
```
`n` is the default value for `bool` symbols. Leaving out the `default n` generates nicer documentation too.

SavinayDharmappa replied:
ok. I will do the change

nvlsianpu commented on include/arch/x86/linker.ld:
```diff
@@ -105,6 +106,15 @@ SECTIONS
 
 	_image_text_end = .;
+
+#ifdef CONFIG_VERSION_HEADER
```
I see the point. For instance bootloader (for zephyr mcuboot) should be as small as possible  so we usually fighting for every byte less in the footprint.

galak replied:
> I see the point. For instance bootloader (for zephyr mcuboot) should be as small as possible so we usually fighting for every byte less in the footprint.

I wouldn't expect the bootloader to use an ELF, but a binary, in which this would presumably not exist.

andyross replied:
FWIW: as implemented above this is going into ROM and will be linked at runtime.

nvlsianpu commented on include/kernel_version.h:
```diff
@@ -30,6 +30,62 @@ extern "C" {
 
+struct version_header {
+	/** Always equal to $B!N */
+	u8_t magic[4];
```
Why this is required? I expect that address of this structure instance will be obtained from the elf or map file.

SavinayDharmappa replied:
It is to uniquely identify the version_header structure.

SavinayDharmappa commented:
@andrewboie @galak  to detect any changes of symbols names in future Zephyr releases in an debug environment.

galak reviewed (changes requested):
If we do support this, I don't think it should be optional.  I assume this is only in the ELF and not in a .bin, so the extra space shouldn't be a concern.

andyross commented on include/arch/x86/linker.ld:
```diff
@@ -105,6 +106,15 @@ SECTIONS
+	SECTION_PROLOGUE (version_header_section, (OPTIONAL),)
+	{
+		*(.version_header)
+		KEEP(*(".version_header*"))
```
1. The second linkage wildcard subsumes the first, making the first one a noop.

2. Why is one KEEP() but the other not?

andyross commented on include/arch/x86/linker.ld:
```diff
@@ -85,6 +85,7 @@ SECTIONS
 #else
 	_image_text_start = .;
 #endif
+
```
I don't understand why this is happening in the x86 linker script.  Can't you put it in common-rom.ld so all architectures can use it?

andyross reviewed (changes requested):
Would be nicer for this to be cross-architecture and not x86-specific.  Some nitpicking on the linker script syntax.

Also Kumar's point is well-taken: what's the value of putting this in runtime memory vs. a separate ELF segment that doesn't get loaded?

SavinayDharmappa commented:
@nashif  we already have test tests/kernel/common/src/main.c test_version() wouldn't that be suffice.

SavinayDharmappa committed:
test: kernel: version_header: Enable CONFIG_VERSION_HEADER.

patch demonstrate enbling CONFIG_VERSION_HEADER will add
version_header section to the elf.

Signed-off-by: Savinay Dharmappa <savinay.dharmappa@intel.com>

pswarnak commented:
recheck

dbkinder commented on tests/kernel/version_header/README.rst:
```diff
@@ -0,0 +1,29 @@
+
+Overview
+********
+A simple program which add the version header to elf when built
```
This should be a complete sentence:

    This program test that the version header is added to the 
    elf binary when the :option:`CONFIG_VERSION_HEADER`
    option is enabled.

SavinayDharmappa replied:
ok.

dbkinder commented on tests/kernel/version_header/src/main.c:
```diff
@@ -0,0 +1,29 @@
+/*
+ * Copyright (c) 2012-2014 Wind River Systems, Inc.
```
Did someone at Wind River write this test?

SavinayDharmappa replied:
ok

SavinayDharmappa commented:
@galak @andyross  version header will be part of both binary and ELF. The idea is to have version_header structure object as part of ELF and binary ,even though this object is not referenced by the code(override linker optimization).  This structure object information can be filled externally as this object is part of special section.

SebastianBoe commented on cmake/host-tools-zephyr.cmake:
```diff
@@ -15,7 +15,7 @@ if(NOT ZEPHYR_SDK_INSTALL_DIR)
   return()
 endif()
 
-set(REQUIRED_SDK_VER 0.9.3)
```
Why are we changing the SDK version?

SavinayDharmappa replied:
my bad..will remove it

SebastianBoe commented on include/kernel_version.h:
```diff
@@ -30,6 +30,62 @@ extern "C" {
 #define SYS_KERNEL_VER_MINOR(ver) (((ver) >> 16) & 0xFF)
 #define SYS_KERNEL_VER_PATCHLEVEL(ver) (((ver) >> 8) & 0xFF)
 
+struct version_header {
```
There are no dependencies between the existing code in kernel_version.h/version.c and the new 'version_header' feature.

To cleanly separate the two features this new feature should be placed in it's own header and source pair.

Doing this will allow us to:

- reduce the number of #ifdef's
- Reduce the noise in kernel_version.h for users that do not use CONFIG_VERSION_HEADER
- Make it clear how version_header and sys_kernel_version interact (they don't)

SavinayDharmappa replied:
Ok.

dbkinder reviewed (approved):
+1 for doc changes, thanks.

SavinayDharmappa committed:
scripts: Add python scripts.

patch adds python scripts which add major, minor and
patchlevel information to version header members in
zephyr.bin

Signed-off-by: Savinay Dharmappa <savinay.dharmappa@intel.com>

SavinayDharmappa committed:
misc: add a section to store kernel version information

Add a section to store kernel version information
This feature is enabled based on config CONFIG_VERSION_HEADER.

Signed-off-by: Savinay Dharmappa <savinay.dharmappa@intel.com>

cinlyooi-intel commented on tests/kernel/version_header/src/main.c:
```diff
@@ -0,0 +1,31 @@
+		     KERNEL_VERSION_MINOR, "minor version mismatch");
+	zassert_true(version_header.patch
+		     == KERNEL_PATCHLEVEL, "patchlevel version match");
+#endif
```
#else 
      zassert_unreachable("CONFIG_VERSION_HEADER not defined");

Please consider adding this #else statement. This will ensure compilation that did not add CONFIG_VERSION_HEADER will failed and not slipping through unexpectedly. If the test should not be run on the platform, then it should be excluded in testcase.yaml

cinlyooi-intel reviewed (commented):
I am having problems running  tests/kernel/version_header. 

```
Assertion failed at 
./tests/kernel/version_header/src/main.c:20: test_version: (version_header.major == KERNEL_VERSION_MAJOR ) major version mismatch
```
After I add printk() statement in main.c:
```
printk("EXPECTED=%d.%d.%d\n", KERNEL_VERSION_MAJOR, KERNEL_VERSION_MINOR, KERNEL_PATCHLEVEL);
printk("GOT     =%d.%d.%d\n", version_header.major, version_header.minor, version_header.patch);
```
I get
```
EXPECTED=1.13.99
GOT     =0.0.0
```
I got it on multiple platforms, e.g. 'altera_max10'.

My .config file has CONFIG_VERSION_HEADER. include/generated/version.h contains the KERNEL_VERSION_XXX information which matches the result.

cinlyooi-intel commented:
Can anyone help me on why I failed to get the required version in the test case please?

> I am having problems running tests/kernel/version_header.
> 
> ```
> Assertion failed at 
> ./tests/kernel/version_header/src/main.c:20: test_version: (version_header.major == KERNEL_VERSION_MAJOR ) major version mismatch
> ```
> After I add printk() statement in main.c:
> 
> ```
> printk("EXPECTED=%d.%d.%d\n", KERNEL_VERSION_MAJOR, KERNEL_VERSION_MINOR, KERNEL_PATCHLEVEL);
> printk("GOT     =%d.%d.%d\n", version_header.major, version_header.minor, version_header.patch);
> ```
> I get
> 
> ```
> EXPECTED=1.13.99
> GOT     =0.0.0
> ```
> I got it on multiple platforms, e.g. 'altera_max10'.
> 
> My .config file has CONFIG_VERSION_HEADER. include/generated/version.h contains the KERNEL_VERSION_XXX information which matches the result.

cinlyooi-intel commented on tests/kernel/version_header/README.rst:
```diff
@@ -0,0 +1,34 @@
+.. zephyr-app-commands::
+   :zephyr-app: kernel/version_header
+   :host-os: unix
+   :board: qemu_x86
```
Looking at it from an alternate view point: Why do testcase.yaml exclude all qemu targets? The test case passed on qemu_x86 when I run it.

marc-hb commented on scripts/binary_version_header.py:
```diff
@@ -0,0 +1,109 @@
+VERSION = [0x01,0x02]
+VERSION_OFFSET = 4
+
+_fields_V1 = [
```
Why are there V1 and V2 versions already, has this already used in some product somewhere? (not asking which/where)

marc-hb commented on scripts/binary_version_header.py:
```diff
@@ -0,0 +1,109 @@
+    ("reserved_2", ctypes.c_ubyte * 4)
+]
+
+def ctypes_factory(buf):
```
This seems convoluted, why no `class BinaryVersionHeaderV2(BinaryVersionHeader)` ?

marc-hb commented on tests/kernel/version_header/README.rst:
```diff
@@ -0,0 +1,34 @@
+This program tests that the version header is added to the elf binary
+when the :option:`CONFIG_VERSION_HEADER`  option is enabled:
+
+vesion header members can be modified with script add_binary_version_header.py.
```
typo

marc-hb commented on scripts/add_binary_version_header.py:
```diff
@@ -0,0 +1,83 @@
+#!/usr/bin/python
```
This script doesn't **add** version headers; it replaces them. Rename accordingly.

marc-hb replied:
How about adding a subdirectory like this: `scripts/version_header/replace_version.py` ?

Due to its super vague name, the `scripts/` directory is unsurprisingly becoming a dumpsite of bits totally unrelated to each other.

If these scripts are only for testing then they should even be moved to `tests/kernel/version_header/`

marc-hb commented on include/version_header.h:
```diff
@@ -0,0 +1,68 @@
+
+	/** Filled with zeros, can be eventually used for 64 bits support */
+	u8_t reserved_2[4];
+} __packed;
```
This should be a 100% generic feature yet ___packed is not very portable. If you layout things carefully then you shouldn't need this. http://www.catb.org/esr/structure-packing/

One of the tests should have a full, completely harcoded byte array that you compare against. This will also test the endianness.

marc-hb reviewed (changes requested):
"Add python scripts" is not a commit subject line, no specifics.

andrewboie commented:
Given that many people have objections to this patch and the original author is no longer working on Zephyr, I think we should just close this.

carlescufi commented:
@marc-hb maybe you could pick this up in a new PR
